### PR TITLE
21601/notifications

### PR DESCRIPTION
**Pull request by DLukash**
- added email verification mechanism
- added mechanism to send emails asynchronously 
- added notification mechanism about published results

### CODE REVIEW

File: webapplication/aoi/management/commands/notebook_executor.py [hunk 1]
```diff
@@ -4,8 +4,8 @@
 from aoi.management.commands._notebook import (
     NotebookDockerThread, 
     PublisherThread, 
-    NotebookK8sThread
 )
+from notificator.management.commands.notification_thread import NotificationThread
```
Comment (adolgarev) [inline webapplication/aoi/management/commands/notebook_executor.py:8]: aoi in this way depends on notificator

**Review by adolgarev — COMMENTED** [review]
Lets simplify a bit
* email verification is not needed at this stage
* no per request setting, only per user setting whether to send notification or not

Lets use separate field for notifications in user profile e.g. notification_email, not email field which can be used for authentication purposes.

File: webapplication/aoi/migrations/0026_auto_20230124_1319.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        migrations.AddField(
+            model_name='component',
+            name='scihub_creds_required',
+            field=models.BooleanField(default=False, verbose_name='Copernicus Open Access Hub credentials are required'),
+        ),
+        migrations.AddField(
+            model_name='component',
+            name='sentinel1_aws_creds_required',
+            field=models.BooleanField(default=False, verbose_name='Sentinel 1 AWS credentials are required'),
+        ),
```
[the remark this answers is not in the document]
Comment (DLukash) [reply webapplication/aoi/migrations/0026_auto_20230124_1319.py]: Moved to another PR 
https://github.com/QuantuMobileSoftware/geoap/pull/327

File: webapplication/aoi/management/commands/notebook_executor.py (hunk 1 above)
Comment (DLukash) [reply webapplication/aoi/management/commands/notebook_executor.py]: I used another approach with save imports

**Review by adolgarev — COMMENTED** [review]
The previous review comment still holds.
image

Regarding [decoupled mail sending](https://github.com/QuantuMobileSoftware/geoap/pull/326/commits/e9cf13b1689158a68421c0aa1f542cfb84708182) . I'd prefer the following.
1. The command webapplication/notificator/management/commands/notificator_k8s.py should be run in both cases: docker and k8s (name it accordingly, note that it was a bit of mistake to name commands as nouns like notebook_executor, they should be verbs like migrate, collect_static, etc.)
2. Run command webapplication/notificator/management/commands/notificator_k8s.py from tasks.py the same way notebook_executor is being run.

File: webapplication/notificator/models.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from django.db import models
+from publisher.models import Result
+
+# Create your models here.
+
+class Notification(models.Model):
+
+    result = models.OneToOneField(Result, on_delete=models.deletion.CASCADE, primary_key=True)
```
Comment (adolgarev) [inline webapplication/notificator/models.py:8]: Please notify about request being completed, not each result
Please add FK to user
Please add createdat = models.DateTimeField(default=timezone.now)

File: webapplication/notificator/models.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class Email(models.Model):
+
+    template = models.CharField(max_length=225, verbose_name="Email's template")
+    subject = models.CharField(max_length=225, verbose_name="Email's subject")
+    recipient_email = models.EmailField()
+    context = models.JSONField()
```
Comment (adolgarev) [inline webapplication/notificator/models.py:16]: This is not needed as we already send email in a background process

File: webapplication/notificator/templates/email_notification.html [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Hello {{ username }}, your request of component {{ component }} have been processed.
+Result for timeperiod from {{ startdate }} to {{ enddate }} is published.
```
Comment (adolgarev) [inline webapplication/notificator/templates/email_notification.html:2]: This should include links to all results associated with request. As well request status (failed, completed, etc.)

File: webapplication/notificator/utils.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        subject (str): subject of the letter
+        context (dict): dictionary of data to fill template and make an email
+    """
+    msg = Email(
+        template=template,
+        recipient_email=recipient_email,
+        subject=subject,
+        context=context
+    )
+    msg.save()
```
Comment (adolgarev) [inline webapplication/notificator/utils.py:19]: Not needed

File: webapplication/notificator/management/commands/notify_results.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            template="email_notification.html",
+            recipient_email=result.request.user.email,
+            subject="Result is awaiable",
+            context={
+                    "username":result.request.user.get_full_name(),
+                    "component":result.request.component.name,
+                    "startdate":result.request.date_from.strftime("%Y-%m-%d"),
+                    "enddate":result.request.date_to.strftime("%Y-%m-%d")
+                }
+        )
```
Comment (adolgarev) [inline webapplication/notificator/management/commands/notify_results.py:44]: Send emails here via django https://docs.djangoproject.com/en/4.1/topics/email/

File: webapplication/notificator/management/commands/send_emails.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+                    email.template,
+                    email.context
+                ),
+                recipient_list=[email.recipient_email],
+                from_email=None
+            )
+            email.delete()
+        logger.info("All emails have been sent succesfuly")
+        
+
```
Comment (adolgarev) [inline webapplication/notificator/management/commands/send_emails.py:38]: This is not needed

**Pull request closed without merging**